matrix/src/armory/matrix/matrix.py

```diff
@@ -0,0 +1,267 @@
+"""Matrix generation utilities"""
+
+import concurrent.futures
+from copy import deepcopy
+from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, Tuple
+
+
+def product(
+    prior: Dict[str, Any], remaining: List[Tuple[str, Any]]
+) -> Iterable[Mapping[str, Any]]:
+    """
+    Generates a cartesian product of parameters.
+
+    This function works recursively, iterating over each parameter range in a
+    depth-first manner. We do this rather than use `itertools.product` in order
+    to support dynamic, or dependent, parameter ranges that are created based
+    on the preceding parameter values.
+
+    Args:
+        prior: Key-value parameter mapping of parameters generated so far
+        remaining: List of remaining key and value-iterable pairs
+
+    Yields:
+        Key-value parameter mappings for each row in the cartesian product
+        matrix
+    """
+    if not remaining:
+        yield prior
+    else:
+        key, values = remaining[0]
+        if callable(values):
+            values = values(**prior)
+        if not isinstance(values, Iterable):
+            values = [values]
+        for value in values:
+            next = deepcopy(prior)
+            next[key] = value
+            for params in product(next, remaining[1:]):
+                yield params
+
+
+def create_matrix(
+    worker_num: Optional[int] = None,
```

Comment:
Naming the partitioning parameters `worker_num` and `num_workers` makes it look like you are mixing in a computational model with the representation of the parameter space. Maybe use `partition_num` and `num_partitions`?

Reply:
Originally I was going to have it automatically partition based on slurm environment variables, but I've decided against that and just made it part of the API. But the variable names never got updated...